crates/red_knot_python_semantic/src/semantic_index/use_def/symbol_state.rs

```diff
@@ -75,55 +81,87 @@ const INLINE_CONSTRAINT_BLOCKS: usize = 2;
 /// Can keep inline this many live bindings per symbol at a given time; more will go to heap.
 const INLINE_BINDINGS_PER_SYMBOL: usize = 4;
 
-/// One [`BitSet`] of applicable [`ScopedConstraintId`] per live binding.
-type InlineConstraintArray = [BitSet<INLINE_CONSTRAINT_BLOCKS>; INLINE_BINDINGS_PER_SYMBOL];
-type Constraints = SmallVec<InlineConstraintArray>;
-type ConstraintsIterator<'a> = std::slice::Iter<'a, BitSet<INLINE_CONSTRAINT_BLOCKS>>;
+/// Which constraints apply to a given binding?
+type Constraints = BitSet<INLINE_CONSTRAINT_BLOCKS>;
+
+type InlineConstraintArray = [Constraints; INLINE_BINDINGS_PER_SYMBOL];
+
+/// One [`BitSet`] of applicable [`ScopedConstraintId`]s per live binding.
+type ConstraintsPerBinding = SmallVec<InlineConstraintArray>;
+
+/// Iterate over all constraints for a single binding.
+type ConstraintsIterator<'a> = std::slice::Iter<'a, Constraints>;
 type ConstraintsIntoIterator = smallvec::IntoIter<InlineConstraintArray>;
 
+/// Similar to what we have above, but for visibility constraints.
+#[newtype_index]
+pub(crate) struct ScopedVisibilityConstraintId;
+
+impl ScopedVisibilityConstraintId {
+    pub(crate) const ALWAYS_TRUE: ScopedVisibilityConstraintId =
+        ScopedVisibilityConstraintId::from_u32(0);
+}
+
+const INLINE_VISIBILITY_CONSTRAINTS: usize = 4;
+type InlineVisibilityConstraintsArray =
+    [ScopedVisibilityConstraintId; INLINE_VISIBILITY_CONSTRAINTS];
+type VisibilityConstraintPerBinding = SmallVec<InlineVisibilityConstraintsArray>;
```

Comment:
Ah, I had missed the singular/plural difference! I agree that "a visibility constraint can be a tree of visibility constraints" is not a good reason to use plural here.

I think your two-identical-types solution here is fine, but I also think that since we've already lost direct parallel with `ConstraintsPerBinding`, it would be fine to lose parallel here as well, and have `ConstraintsPerBinding` and `VisibilityConstraintsPerDefinition`. This also represents a real difference in how they are used.

Mostly I think it would be good in your documentation pass to add some comments on these types; right now the only comment says "similar to above but for visibility constraints", when actually there are key differences from above.